source/common/upstream/load_balancer_impl.cc

```diff
@@ -88,6 +99,11 @@ void LoadBalancerBase::recalculatePerPriorityState(uint32_t priority) {
     per_priority_load_[i] =
         std::min<uint32_t>(total_load, per_priority_health_[i] * 100 / total_health);
     total_load -= per_priority_load_[i];
+
+    // For each level check if it should run in Panic mode. Never set Panic mode if the total health
+    // is 100%, even when individual priority level has very low # of healthy hosts.
```

Comment:
This plays interestingly with things like #4529 so I just want it on @snowp 's radar.  

For simplicity's sake I'd prefer that is if we're only temporarily excluding a priority level due to retries we not recalculate per_priority_panic_ based on if we'd turn up panic mode if that priority were truly not in play.  Realistically if we're doing retries it's probably because things we hope were healthy are not, and we if it's bad enough  we're failing over we may also want to lean in the direction of panic mode.  I'll leave that up to Snow :-P

Reply:
As #4529 stands this won't be affected by the excluded priorities since I rebalance based on those exclusions *after* calling `recalculatePerPriorityState`. We'd be using the priority load adjusted for panic mode and redistribute the load of excluded priorities between the ones not excluded, which I think is a sane behavior. You'd honor panic mode but make an extra effort to spread retries between distinct priorities.

tl;dr I think this is fine as is

Reply:
I was ready for another code push but did rebase, because github complained about conflict. After @snowp changes for #4529, `recalculatePerPriorityState` is static method now. In this PR, I added call to `isGlobalPanic` which is not static and requires LoadBalancerBase object to access configuration for panic thresholds. Probably I have to calculate panic states outside of `recalculatePerPriorityState` or pass more params to it, which is bit ugly.

Reply:
I think adding it as a parameter isn't that bad. If the parameter list grows too big maybe we should add something like a `PerPriorityLoadBalancerState` struct to encapsulate all the per priority state